### PR TITLE
persist: wire up coordinator to allow compaction in persistence

### PR DESCRIPTION
**Pull request by ruchirK**
I think this is a reasonable first cut implementation for wiring up the coordinator to send allow compaction messages aka since frontier updates to the persistence crate.

Normally, since frontiers are derived from the upper frontier for the same object and the set of transactions, sinks and other interlopers who wish to hold the since frontier at a particular time. However, tables don't yet have a notion of upper frontiers so porting the standard approach used by indexes and sources is nontrivial. Instead, this commit teaches the coordinator to specifically inform the persistence crate of the table's default index's since frontier whenever it is updated. This approach is safe, because the default index can never be deleted unless the table itself is deleted, and there is never more than one default index. Additionally, this approach introduces no new state in the coordinator.

However, this approach introduces a small runtime cost in `maintenance` where we now iterate through all of the received since frontier updates and try to see if they came from an index backing a persisted table.

### CODE REVIEW

**Comment by ruchirK** [comment]
RFAL!

@mjibson you'll be happy to know that coordtest found a bug in my first attempt that nothing else found (I searched the catalog for all ids with since updates, forgetting about transient ids). I also found a bug in coordtest -- it hangs forever when the underlying coordinator panics :D

File: src/coord/src/coord.rs
Comment (mjibson) [inline src/coord/src/coord.rs:1203]: Should this go into the not empty check block below? No idea if it'd result in meaningful benefit.

File: src/coord/src/coord.rs
Comment (danhhz) [inline src/coord/src/coord.rs:1209]: it strikes me as a little odd that one of these uses try_get_by_id and other other get_by_id. I'd say either standardize on one of leave a comment on why they're different

File: src/coord/src/coord.rs
Comment (danhhz) [inline src/coord/src/coord.rs:1203]: let's moves this out into a helper function

File: src/coord/src/coord.rs
Comment (danhhz) [inline src/coord/src/coord.rs:1212]: let's rebase this on top of my recent PR and resolve this TODO (and also use multiwriteahandle to make this into one persist call per maintenance call)

File: src/coord/src/coord.rs
Comment (mjibson) [inline src/coord/src/coord.rs:1214]: If it doesn't affect correctness we can probably ignore the response?

File: src/coord/src/coord.rs
Comment (mjibson) [inline src/coord/src/coord.rs:1212]: Can you describe this a bit more?

File: src/coord/src/coord.rs
Comment (danhhz) [inline src/coord/src/coord.rs:1214]: Not doing something with the response scares me. It's already proves useful (at the very least to philip) to be printing errors when these things fail for seal (which also doesn't await anything in the critical path), so I think we should do the same here

**Review by mjibson — APPROVED** [review]

**Review by danhhz — APPROVED** [review]
lgtm after addressing these comments

fwiw, benesch signed off on the approach offline, modulo the helper function comment which came from him

I realized another approach we could take is to store a copy of the write handle on each Index item, in parallel to where it is for tables, given that we make them in pairs, but i'm not sure that's actually any better. i lean toward just merging this one to unblock testing and we can revisit that idea if someone falls in love with it

File: src/coord/src/coord.rs
Comment (mjibson) [reply src/coord/src/coord.rs:1214]: What could we do that's useful? Printing a log message seems not beneficial. This isn't in a user path so there's nothing to tell them. It doesn't affect correctness to ignore compaction.

File: src/coord/src/coord.rs
Comment (mjibson) [reply src/coord/src/coord.rs:1209]: I think the current stuff is correct (the index might not exist, but if it does its table must).

File: src/coord/src/coord.rs
Comment (ruchirK) [reply src/coord/src/coord.rs:1214]: I don't have a strong feeling. I just noted that this was the precedent set by the coordinator in how it gives since updates to the dataflow workers. Happy to defer to whatever you both decide on.

File: src/coord/src/coord.rs
Comment (ruchirK) [reply src/coord/src/coord.rs:1209]: ill add the comment but thats the reasoning yup

File: src/coord/src/coord.rs
Comment (danhhz) [reply src/coord/src/coord.rs:1214]: I'm not sure that's a relevant precedent. The insert, seal, and allow_compaction messages that coord sends to dataflow are all infallible, but the persist equivalents are not and that's a pretty big difference.

I've been trained for years and years to never ignore errors and it's served me well. Logging this one isn't that actionable by a user (maybe? if they tell us about these errors that's actually a good outcome), but it's better than silently ignoring it. We could have a counter for failures, which might be the right answer in the long run, but in the short term I think that's even less likely to get looked at than logs.

> It doesn't affect correctness to ignore compaction.

Possibly nitpicky, but I'm not sure I agree with this. Ignoring an individual compaction message doesn't affect correctness, agreed, but we "guarantee" (intend to guarantee) an amount of disk usage somehow proportional to the size of the non-retracted records in a collection. If we ignore sustained failure in compaction messages, then we won't be living up to that.

On the other hand, given the indirectness of the id-hopping we're doing here, it doesn't really inspire confidence that we're actually compacting all the tables even if nothing is erroring. Seems like an easy codepath to accidentally break. So perhaps there's some larger thing we need to be doing here to be ensuring that everything is healthy (monitor the compaction frontier of all active persisted collections, for example).

My strong inclination is to make this a log message with a TODO to figure out something smarter. If matt is uncomfortable with that, then just the TODO is fine for now.

Comment (mjibson) [reply src/coord/src/coord.rs]: sgtm

File: src/coord/src/coord.rs
Comment (ruchirK) [reply src/coord/src/coord.rs]: resolved the todo. I did not change this to use multiwritehandle at this time because the api requires a list of (Id, Frontier) pairs and we would need a map from globalid -> id on PersistMultiDetails (or something like that) to make this work if I understand correctly.

It also seemed to me like the perf win from batching up multiple allow_compaction messages into one wouldn't be that big even with many tables because the batched message does essentially the same work as a run of unbatched messages (modify the in memory state and wait to get flushed by drain_pending) and the batched message itself is not dramatically smaller (unlike the seal api where we can get away with a list of ids and a single seal frontier). In this case, the batched message is likely approximately the same size as the sum of the unbatched message sizes minus some constant overhead.

Given that plus the additional complexity of changing PersistMultiWrites I felt it was ok to punt on the multi-write-handle work for allow_compaction. wdyt?

Comment (danhhz) [reply src/coord/src/coord.rs]: I'd rather do this now. You should be able to call `persist.write_handle.stream_id()` to get the persist `Id`

Comment (ruchirK) [reply src/coord/src/coord.rs]: aha! forgot about that function. yup done

**Comment by ruchirK** [comment]
tftr! merging on green!

**Pull request merged**
